File: Projects/FlutterPerfLab/lib/modules/sensortest/sensor_test_view_model.dart
import 'dart:async';
import 'package:flutter/foundation.dart';
import 'package:share_plus/share_plus.dart';

import '../../core/managers/performance_manager.dart';
import '../../core/managers/sensor_test_manager.dart';
import '../../core/managers/export_manager.dart';

/// A lifecycle-aware ViewModel orchestrating high-frequency (100Hz) sensor telemetry.
///
/// Employs a pure monotonic Stopwatch to guarantee absolute temporal precision
/// and prevent timer drift during the benchmark.
class SensorTestViewModel extends ChangeNotifier {
  // MARK: - Reactive UI State
  bool isRunning = false;
  String status = "Sensors Ready";
  bool isReportReady = false;
  double progress = 0.0;

  // MARK: - Hoisted Telemetry State
  // Primitive arrays optimized for low-memory representation.
  List<double> accelData = [0.0, 0.0, 0.0];
  List<double> gyroData = [0.0, 0.0, 0.0];
  List<double> magnetData = [0.0, 0.0, 0.0];
  int stepCount = 0;

  // MARK: - Manager Integration
  final PerformanceManager _performance = PerformanceManager.shared;
  final SensorTestManager _sensorManager = SensorTestManager.shared;

  // Standardized benchmark duration (60 seconds) strictly enforced.
  final double _testDuration = 60.0;

  // Concurrency & Timing Management
  Timer? _timerTask;
  final Stopwatch _hardwareClock = Stopwatch();

  // MARK: - Benchmark Lifecycle

  void startTest() {
    if (isRunning) return;

    isRunning = true;
    status = "Acquiring Telemetry (100Hz)...";
    isReportReady = false;
    progress = 0.0;

    _performance.startMonitoring();

    // Directly invokes the Singleton engine to prevent allocation overhead.
    _sensorManager.startSensors(
      accelCallback: (data) {
        accelData = data;
        notifyListeners();
      },
      gyroCallback: (data) {
        gyroData = data;
        notifyListeners();
      },
      magnetCallback: (data) {
        magnetData = data;
        notifyListeners();
      },
      stepCallback: (count) {
        stepCount = count;
        notifyListeners();
      },
    );

    // Captures a monotonic baseline to eliminate timer drift.
    _hardwareClock.reset();
    _hardwareClock.start();

    // Cooperative event loop polling with a 100ms delay.
    _timerTask = Timer.periodic(const Duration(milliseconds: 100), (_) {
      _updateProgress();
    });

    notifyListeners();
  }

  void stopTest({bool isFinished = false}) {
    if (!isRunning) return;

    isRunning = false;
    status = isFinished ? "✅ Test Finalized" : "🛑 Terminated";

    // Gracefully cancel the asynchronous timing task.
    _timerTask?.cancel();
    _timerTask = null;
    _hardwareClock.stop();

    // Hardware Teardown
    _sensorManager.stopSensors();

    _performance.stopMonitoring();

    // Flush stale telemetry data to ensure visual cleanliness and GC release.
    accelData = [0.0, 0.0, 0.0];
    gyroData = [0.0, 0.0, 0.0];
    magnetData = [0.0, 0.0, 0.0];
    stepCount = 0;

    if (isFinished) {
      progress = 1.0;
      isReportReady = true;
    } else {
      progress = 0.0;
      isReportReady = false;
    }

    notifyListeners();
  }

  // Exports telemetry via the native share sheet.
  Future<void> exportResults() async {
    try {
      final file = await ExportManager.shared.generateCSV(
        logs: _performance.currentLogs,
        testName: "Sensor_Flutter_Stress_100Hz",
        customSummary: "Total_Steps,$stepCount,,,,",
      );

      if (file != null) {
        final xFile = XFile(file.path);
        await Share.shareXFiles([xFile]);
        status = "✅ Export Ready.";
      } else {
        status = "❌ CSV Export Failed (Empty logs)";
      }
    } catch (e) {
      status = "❌ CSV Export Failed: $e";
    }
    notifyListeners();
  }

  // MARK: - Internal Engine

  void _updateProgress() {
    // Always reads from the monotonic clock to compensate for timer inaccuracies.
    final double elapsedSeconds = _hardwareClock.elapsedMilliseconds / 1000.0;

    progress = (elapsedSeconds / _testDuration).clamp(0.0, 1.0);

    if (elapsedSeconds >= _testDuration) {
      stopTest(isFinished: true);
    } else {
      notifyListeners();
    }
  }

  @override
  void dispose() {
    stopTest();
    super.dispose();
  }
}

File: Projects/KotlinMultiPlatformPerfLab/composeApp/src/commonMain/kotlin/com/tez/perflab/ui/features/sensortest/SensorTestViewModel.kt
package com.tez.perflab.ui.features.sensortest

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.tez.perflab.managers.ExportManager
import com.tez.perflab.managers.PerformanceManager
import com.tez.perflab.managers.PlatformContext
import com.tez.perflab.managers.SensorTestManager
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlin.time.Duration.Companion.seconds
import kotlin.time.TimeSource

/**
 * A lifecycle-aware orchestrator managing high-frequency (100Hz) sensor telemetry.
 */
class SensorTestViewModel : ViewModel() {

    // MARK: - Reactive UI State

    private val _isRunning = MutableStateFlow(false)
    val isRunning: StateFlow<Boolean> = _isRunning.asStateFlow()

    private val _status = MutableStateFlow("Sensors Ready")
    val status: StateFlow<String> = _status.asStateFlow()

    private val _progress = MutableStateFlow(0.0)
    val progress: StateFlow<Double> = _progress.asStateFlow()

    private val _isReportReady = MutableStateFlow(false)
    val isReportReady: StateFlow<Boolean> = _isReportReady.asStateFlow()

    // ─── HOISTED TELEMETRY STATE ──────────────────────────────────────────────
    // StateFlow observables are hoisted from the hardware manager directly into the
    // ViewModel. This aligns memory management with native baselines,
    // ensuring identical Garbage Collection (GC) footprints during benchmarks.

    private val _accelData = MutableStateFlow(floatArrayOf(0f, 0f, 0f))
    val accelData: StateFlow<FloatArray> = _accelData.asStateFlow()

    private val _gyroData = MutableStateFlow(floatArrayOf(0f, 0f, 0f))
    val gyroData: StateFlow<FloatArray> = _gyroData.asStateFlow()

    private val _magnetData = MutableStateFlow(floatArrayOf(0f, 0f, 0f))
    val magnetData: StateFlow<FloatArray> = _magnetData.asStateFlow()

    private val _stepCount = MutableStateFlow(0)
    val stepCount: StateFlow<Int> = _stepCount.asStateFlow()

    // MARK: - Benchmark Execution State

    private val testDuration = 60.seconds
    private var timerJob: Job? = null

    fun startTest(context: PlatformContext) {
        if (_isRunning.value) return
        stopTest(isFinished = false)

        _isRunning.value = true
        _isReportReady.value = false
        _status.value = "Acquiring Telemetry (100Hz)..."
        _progress.value = 0.0

        PerformanceManager.startMonitoring(context)

        // Direct injection of state-mutating lambdas into the hardware engine.
        SensorTestManager.startSensors(
            context = context,
            onAccelUpdate = { _accelData.value = it },
            onGyroUpdate = { _gyroData.value = it },
            onMagnetUpdate = { _magnetData.value = it },
            onStepUpdate = { _stepCount.value = it }
        )

        // Drift-Free Isolated Progress Tracking
        timerJob = viewModelScope.launch(Dispatchers.Default) {
            val timeMark = TimeSource.Monotonic.markNow()

            while (isActive && timeMark.elapsedNow() < testDuration) {
                delay(100)
                val elapsed = timeMark.elapsedNow()
                _progress.value = (elapsed / testDuration).coerceIn(0.0, 1.0)
            }

            if (_isRunning.value) {
                stopTest(isFinished = true)
            }
        }
    }

    fun stopTest(isFinished: Boolean = false) {
        if (!_isRunning.value) return

        _isRunning.value = false
        _status.value = if (isFinished) "✅ Test Finalized" else "🛑 Terminated"

        // Hardware Teardown
        timerJob?.cancel()
        SensorTestManager.stopSensors()
        PerformanceManager.stopMonitoring()

        // Flush stale telemetry data to clear the UI graphs.
        _accelData.value = floatArrayOf(0f, 0f, 0f)
        _gyroData.value = floatArrayOf(0f, 0f, 0f)
        _magnetData.value = floatArrayOf(0f, 0f, 0f)
        _stepCount.value = 0

        if (isFinished) {
            _progress.value = 1.0
            _isReportReady.value = true
        }
    }

    fun shareResults(context: PlatformContext) {
        viewModelScope.launch(Dispatchers.Default) {
            try {
                ExportManager.generateAndShareCSV(
                    logs = PerformanceManager.currentLogs.value,
                    testName = "Sensor_KMP_Stress_100Hz",
                    context = context
                )
            } catch (e: Exception) {
                _status.value = "❌ Export Failed: ${e.message}"
            }
        }
    }

    override fun onCleared() {
        super.onCleared()
        timerJob?.cancel()
        SensorTestManager.stopSensors()
        PerformanceManager.stopMonitoring()
    }
}